### ISSUE TITLE
Error: response code 500 while fetching metrics from Cost explorer

### ISSUE THREAD

**Issue by Alok650** [electrolux-oss/infrawallet#7]
Hi guys,
I tried using the infra wallet plugin for cost explorer metrics on backstage. I have tried integrating with multiple aws accounts. On running the plugin, I am getting error code 500 from /reports api endpoint.

```
{
    "error": {
        "name": "Error",
        "message": "An internal error occurred logId=e0721b098d4b588b14ba"
    },
    "request": {
        "method": "GET",
        "url": "/reports?&filters=&groups=&granularity=monthly&startTime=1711909800000&endTime=1719772199999"
    },
    "response": {
        "statusCode": 500
    }
}
```

I have integrated 2 aws accounts with this plugin i.e. my app-config.yaml:
```
  infraWallet:
    integrations:
      aws:
      - name: accountA
        accountId: ${AWS_ACCOUNT_ID_A}
        roleArn: ${AWS_IAM_ROLE_ARN_A{
        region: ${AWS_REGION}
      - name: accountB
        accountId: ${AWS_ACCOUNT_ID_B}
        roleArn: ${AWS_IAM_ROLE_ARN_B}
        region: ${AWS_REGION}
 ```
 
 Can anyone help with what's wrong or how to debug this?

**Comment by Alok650** [comment]
Further adding logs from app:

```
[1] 2024-06-12T17:56:33.137Z backstage error Filtered internal error with logId=29809580ea7d73ebfb35 from response select * from "category_mappings" where "provider" = $1 - relation "category_mappings" does not exist type=errorHandler logId=29809580ea7d73ebfb35 length=116 name=error severity=ERROR code=42P01 detail=undefined hint=undefined position=15 internalPosition=undefined internalQuery=undefined where=undefined schema=undefined table=undefined column=undefined dataType=undefined constraint=undefined file=parse_relation.c line=1395 routine=parserOpenTable stack=error: select * from "category_mappings" where "provider" = $1 - relation "category_mappings" does not exist
[1]     at Parser.parseErrorMessage (/Users/pixl/Documents/workspace/duplicate/backstage/node_modules/pg-protocol/src/parser.ts:369:69)
[1]     at Parser.handlePacket (/Users/pixl/Documents/workspace/duplicate/backstage/node_modules/pg-protocol/src/parser.ts:188:21)
[1]     at Parser.parse (/Users/pixl/Documents/workspace/duplicate/backstage/node_modules/pg-protocol/src/parser.ts:103:30)
[1]     at Socket.<anonymous> (/Users/pixl/Documents/workspace/duplicate/backstage/node_modules/pg-protocol/src/index.ts:7:48)
[1]     at Socket.emit (node:events:517:28)
[1]     at addChunk (node:internal/streams/readable:368:12)
[1]     at readableAddChunk (node:internal/streams/readable:341:9)
[1]     at Readable.push (node:internal/streams/readable:278:10)
[1]     at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
```

**Comment by emillg** [comment]
@Alok650 At first glance, it looks like your config doesn't follow the README documentation.

Can you update it to below format, and post the new results here?

```
backend:
  infraWallet:
    integrations:
      aws:
        - name: <unique_name_of_this_account>
          accountId: '<12-digit_account_ID>' # quoted as a string
          assumedRoleName: <name_of_the_AWS_IAM_role_to_be_assumed>
          accessKeyId: <access_key_ID_of_AWS_IAM_user_that_assumes_the_role> # optional, only needed when an IAM user is used to assume the role
          accessKeySecret: <access_key_secret_of_AWS_IAM_user_that_assumes_the_role> # optional, only needed when an IAM user is used to assume the role
```

**Comment by Alok650** [comment]
```
  infraWallet:
    integrations:
      aws:
      - name: account_a
        accountId: '<account_id_a>'
        assumedRoleName: backstage-iam
      - name: account_b
        accountId: '<account_id_b>'
        assumedRoleName: backstage-iam
```

still getting 5xx.

**Comment by gluckzhang** [comment]
Hi @Alok650, based on the backend error logs you provided:

```
... "category_mappings" does not exist
```

It seems related to the database. When InfraWallet is added, it first executes a migration script to create a table named `category_mappings` and add a default mapping to the table.

Which database are you using? Is the migration script skipped in your Backstage instance?

**Comment by Alok650** [comment]
I am using postgres. I can't find a separate db for this plugin created. It works for other plugins. Are there any manual steps related to db?

**Comment by gluckzhang** [comment]
> I am using postgres. I can't find a separate db for this plugin created. It works for other plugins. Are there any manual steps related to db?

Postgres should be good. We use Backstage's backend database service (https://backstage.io/docs/tutorials/configuring-plugin-databases/#postgresql) to create the plugin database. So usually you do not need to configure anything. Could you also tell me if you have anything like the following in your config?

```yaml
backend:
  database:
    migration:
      skip: true
```

**Comment by Alok650** [comment]
Hi no,
my db configs are:
```
  database:
    client: pg
    connection:
      host: ${POSTGRES_HOST}
      port: ${POSTGRES_PORT}
      user: ${POSTGRES_USER}
      password: ${POSTGRES_PASSWORD}
    knexConfig:
      pool:
        min: 30
        max: 120
        acquireTimeoutMillis: 60000
        createTimeoutMillis: 30000
        destroyTimeoutMillis: 5000
        idleTimeoutMillis: 60000
        reapIntervalMillis: 1000
        createRetryIntervalMillis: 200
        propagateCreateError: false
    cache:
      store: redis
      connection: ${REDIS_ENDPOINT}
```

**Comment by gluckzhang** [comment]
hmm, OK, for now I cannot reproduce the issue using a local postgres db. The plugin creates a database named `backstage_plugin_infrawallet` and there is a table `category_mappings` created. Maybe you can also check your database directly and see if there are any Backstage plugin databases. This may help me locate the issue, at least to verify whether it is a failure to create database and the table, or a failure to connect/query the data afterward.

**Comment by Alok650** [comment]
@gluckzhang the tables are not getting created:
<img width="321" alt="image" src="https://github.com/electrolux-oss/infrawallet/assets/58817460/a4e55999-e6d5-4b79-93ae-b95ea9b8e078">
the db does get created

**Comment by gluckzhang** [comment]
@Alok650 got it, thanks. I will check the issue further and get back to you soon!

**Comment by gluckzhang** [comment]
Finally, I can reproduce the issue and I think it is related to creating the plugin in a legacy backend system. I will fix the issue soon. Sorry about the inconvenience!

**Comment by gluckzhang** [comment]
Hi @Alok650, we have fixed the issue and released a new version (`0.1.2`) of infrawallet-backend: https://www.npmjs.com/package/@electrolux-oss/plugin-infrawallet-backend

Could you try the following steps and see if it works in your environment?
1. fix the code in attaching the backend, there was a mistake in the doc before, see the latest code here: https://github.com/electrolux-oss/infrawallet?tab=readme-ov-file#if-the-legacy-backstage-backend-system-is-used (`audit: env.audit,` should be `cache: env.cache.getClient(),`)
2. check the `package.json` file in your Backstage instance and make sure to use the latest version of the backend (i.e., `"@electrolux-oss/plugin-infrawallet-backend": "^0.1.2",`)

**Comment by Alok650** [comment]
Hi @gluckzhang. Thanks for the update. its fixed in the latest version.
